File: clojure/day06/index2.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))
(defn split [p s] (s/split s p))

(def data (->> "./input"
  slurp
  s/trim
  (split #"\t")
  (map parse-int)
  (into [])
))

(defn get-max-index* [max max-index index a] (
  if (== 0 (count a)) max-index (
    let [
      e (first a)
      a' (rest a)
    ] (if (> e max)
      (recur e index (inc index) a')
      (recur max max-index (inc index) a'))
  )
))

(defn get-max-index [a] (get-max-index* 0 0 0 a))

(defn nxt [i a] (mod (inc i) (count a)))

(defn redistribute* [rem index a] (
  if (== 0 rem) a (
    recur (dec rem) (nxt index a) (update a index inc)
  )
))

(defn redistribute [a] (
  let [
    i (get-max-index a)
    rem (get a i)
    a' (assoc a i 0)
  ] (redistribute* rem (nxt i a) a')
))

(def period (loop [
  hash {}
  iter 0
  data data
] (let [
  j (get hash data)
] (if j (- iter j) (
    let [
      hash' (assoc hash data iter)
      data' (redistribute data)
    ] (recur hash' (inc iter) data')
  ))
)))

(println period)

File: nodejs/day06/index2.js
const { readLines, parseInt } = require("../lib");

const data = readLines("./input")[0]
  .split("\t")
  .map(parseInt);

const getMaxIndex = () => {
  const max = Math.max(...data);
  return Math.min(
    ...data.map((el, i) => (el === max ? i : null)).filter(el => el != null)
  );
};

const l = data.length;

const next = i => (i + 1) % l;

const redistribute = () => {
  let i = getMaxIndex(),
    b = data[i];

  data[i] = 0;
  while (b--) {
    i = next(i);
    data[i]++;
  }
};

const key = () => data.join("-");

const count = () => {
  const hash = {};
  let iter = 0;

  hash[key()] = 0;
  while (true) {
    redistribute();
    iter++;
    const k = key();
    if (k in hash) {
      return iter - hash[k];
    }
    hash[k] = iter;
  }
};

console.log(count());
